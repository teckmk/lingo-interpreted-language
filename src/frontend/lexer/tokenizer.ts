/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { InvalidIndentationError } from "../../error"
import { Spec, TokenType } from "./specs"

export type Token = {
  type: TokenType
  value: any
  column: number
  line: number
  code?: string
}

export default class Tokenizer {
  private MAX_QUEUE_LEN = 3

  private _spec: Spec[]
  private _tokens: Token[] = []
  private _tokenTypesQueue: TokenType[] = []

  private _cursor = 0
  private _line = 1 // current line
  private _tokenNumber = 1 // current column

  private _code = ""
  private _filename = ""

  private _tabSize = 0
  private _indentation = 0
  private _isIndentMode = false

  constructor(spec: Spec[], filename: string) {
    this._spec = spec
    this._filename = filename
  }

  tokenize(code: string) {
    this._code = code
    this._cursor = 0

    while (!this._isEOF && this._hasNext) this._readToken()

    this._pushToken({
      type: TokenType.EOF,
      value: "EOF",
      column: this._tokenNumber,
      line: this._line,
    })

    return this._tokens
  }

  private get _isEOF(): boolean {
    if (!this._code) return true

    return this._cursor == this._code.length
  }

  private get _hasNext(): boolean {
    if (!this._code || this._code.length == 0) return false

    return this._cursor < this._code.length
  }

  private _readToken() {
    const code = this._code.slice(this._cursor)

    for (const { regex, tokenType } of this._spec) {
      const tokenValue = this._matched(regex, code)

      // skip iteration if there's no match
      if (tokenValue == null) continue

      if (tokenType == TokenType.EOL) {
        this._line++
        this._tokenNumber = 0
      }

      if (tokenType == TokenType.WhiteSpace) {
        const indent = this._handleIndentation(tokenValue)
        if (indent) return
      }

      if (this._isIndentMode) {
        const dedent = this._handleDedentation(tokenValue, tokenType)
        if (dedent) return
      }

      if (
        tokenType == TokenType.EOL ||
        tokenType == TokenType.SingleLineComment ||
        tokenType == TokenType.WhiteSpace
      ) {
        this._updateQueue(tokenType)
        return
      }

      return this._pushToken({
        type: tokenType,
        value: tokenValue,
        column: this._tokenNumber,
        line: this._line,
        // code,
      })
    }

    throw new Error(
      `Unexpected token '${code[0]}' at ${this._line}:${this._tokenNumber} in ${this._filename}`
    )
  }

  private _matched(regex: RegExp, string: string) {
    const matched = regex.exec(string)

    if (matched == null) return null

    this._cursor += matched[0].length
    this._tokenNumber++

    return matched[0]
  }

  private _updateQueue(tokenType: TokenType) {
    const queueLen = this._tokenTypesQueue.length

    if (queueLen == this.MAX_QUEUE_LEN) this._tokenTypesQueue.shift()

    this._tokenTypesQueue.push(tokenType)
  }

  private _pushToken(token: Token) {
    this._tokens.push(token)
    this._updateQueue(token.type)
  }

  private _handleIndentation(tokenValue: string) {
    const q = this._tokenTypesQueue
    const whitespace = tokenValue.length

    if (q.length != this.MAX_QUEUE_LEN) return false

    // handle indent start
    if (q[1] != TokenType.Colon && q[2] != TokenType.EOL) return false

    // remove ':' token so it doesn't bother us in parser
    if (this._tokens.at(-1)?.type == TokenType.Colon) this._tokens.pop()

    if (this._tabSize == 0) this._tabSize = whitespace

    if (whitespace % this._tabSize != 0) throw InvalidIndentationError(this._tabSize, this._line)

    if (whitespace > this._indentation) {
      this._indentation += this._tabSize
      this._isIndentMode = true

      this._pushToken({
        type: TokenType.Indent,
        value: whitespace,
        column: this._tokenNumber,
        line: this._line,
      })

      return true
    }

    return false
  }

  private _handleDedentation(tokenValue: any, tokenType: TokenType) {
    if (this._tokenTypesQueue[2] != TokenType.EOL) return false

    if (tokenType == TokenType.WhiteSpace) {
      const whitespace = tokenValue.length

      if (whitespace < this._indentation) {
        this._indentation -= this._tabSize

        this._pushToken({
          type: TokenType.Dedent, // TokenType.Dedent
          value: tokenValue,
          column: this._tokenNumber,
          line: this._line,
        })

        return true
      }
    }

    if (this._indentation == this._tabSize) {
      this._indentation = 0
      this._isIndentMode = false

      this._pushToken({
        type: TokenType.Dedent, // TokenType.Dedent
        value: tokenValue,
        column: this._tokenNumber,
        line: this._line,
      })

      return true
    }

    return false
  }
}
