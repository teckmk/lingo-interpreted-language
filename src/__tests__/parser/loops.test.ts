import { tokenize } from "../../frontend/lexer/tokenizer"
import Parser from "../../frontend/parser"

describe("Parser - Loops", () => {
  it("should parse infinite for loop", () => {
    const code = "for { }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForStatement",
          label: undefined,
          loopId: "loop_1",
          body: [],
        },
      ],
    })
  })

  it("should parse for loop with condition", () => {
    const code = "for i < 10 { }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForStatement",
          label: undefined,
          loopId: "loop_1",
          condition: {
            kind: "BinaryExpr",
            operator: "<",
            left: {
              kind: "Identifier",
              symbol: "i",
            },
            right: {
              kind: "NumericLiteral",
              value: 10,
            },
          },
          body: [],
        },
      ],
    })
  })

  it("should parse for loop with initializer, condition and update", () => {
    const code = "for var i = 0; i < 10; i = i + 1 { }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForStatement",
          label: undefined,
          loopId: "loop_1",
          initializer: {
            kind: "VarDeclaration",
            identifier: "i",
            type: undefined,
            value: {
              kind: "NumericLiteral",
              value: 0,
            },
            modifier: "variable",
          },
          condition: {
            kind: "BinaryExpr",
            operator: "<",
            left: {
              kind: "Identifier",
              symbol: "i",
            },
            right: {
              kind: "NumericLiteral",
              value: 10,
            },
          },
          update: {
            kind: "AssignmentExpr",
            assigne: {
              kind: "Identifier",
              symbol: "i",
            },
            value: {
              kind: "BinaryExpr",
              operator: "+",
              left: {
                kind: "Identifier",
                symbol: "i",
              },
              right: {
                kind: "NumericLiteral",
                value: 1,
              },
            },
          },
          body: [],
        },
      ],
    })
  })

  it("should parse for in loop", () => {
    const code = "for var i,v in arr { }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForInStatement",
          loopId: "loop_1",
          label: undefined,
          valueIdentifier: "v",
          indexIdentifier: "i",
          iterable: {
            kind: "Identifier",
            symbol: "arr",
          },
          body: [],
        },
      ],
    })
  })

  it("should parse for range loop", () => {
    const code = "for var i,v in range 0 to 10 { }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForRangeStatement",
          loopId: "loop_1",
          label: undefined,
          valueIdentifier: "v",
          indexIdentifier: "i",
          start: {
            kind: "NumericLiteral",
            value: 0,
          },
          end: {
            kind: "NumericLiteral",
            value: 10,
          },
          inclusive: false,
          body: [],
        },
      ],
    })
  })

  it("should parse for range loop with step", () => {
    const code = "for var i,v in range 0 through 10 step 2 { }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForRangeStatement",
          loopId: "loop_1",
          label: undefined,
          valueIdentifier: "v",
          indexIdentifier: "i",
          start: {
            kind: "NumericLiteral",
            value: 0,
          },
          end: {
            kind: "NumericLiteral",
            value: 10,
          },
          step: {
            kind: "NumericLiteral",
            value: 2,
          },
          inclusive: true,
          body: [],
        },
      ],
    })
  })

  it("should parse for loop with label", () => {
    const code = "for label loop { }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForStatement",
          label: "loop",
          loopId: "loop_1",
          body: [],
        },
      ],
    })
  })

  it("should parse for loop with label and condition", () => {
    const code = "for i < 10 label loop { }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForStatement",
          label: "loop",
          loopId: "loop_1",
          condition: {
            kind: "BinaryExpr",
            operator: "<",
            left: {
              kind: "Identifier",
              symbol: "i",
            },
            right: {
              kind: "NumericLiteral",
              value: 10,
            },
          },
          body: [],
        },
      ],
    })
  })

  it("should parse for loop with label, initializer, condition and update", () => {
    const code = "for var i = 0; i < 10; i = i + 1 label loop { }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForStatement",
          label: "loop",
          loopId: "loop_1",
          initializer: {
            kind: "VarDeclaration",
            identifier: "i",
            type: undefined,
            value: {
              kind: "NumericLiteral",
              value: 0,
            },
            modifier: "variable",
          },
          condition: {
            kind: "BinaryExpr",
            operator: "<",
            left: {
              kind: "Identifier",
              symbol: "i",
            },
            right: {
              kind: "NumericLiteral",
              value: 10,
            },
          },
          update: {
            kind: "AssignmentExpr",
            assigne: {
              kind: "Identifier",
              symbol: "i",
            },
            value: {
              kind: "BinaryExpr",
              operator: "+",
              left: {
                kind: "Identifier",
                symbol: "i",
              },
              right: {
                kind: "NumericLiteral",
                value: 1,
              },
            },
          },
          body: [],
        },
      ],
    })
  })

  it("should parse for in range loop with label", () => {
    const code = "for var i,v in range 0 to 10 label loop { }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForRangeStatement",
          loopId: "loop_1",
          label: "loop",
          valueIdentifier: "v",
          indexIdentifier: "i",
          start: {
            kind: "NumericLiteral",
            value: 0,
          },
          end: {
            kind: "NumericLiteral",
            value: 10,
          },
          inclusive: false,
          body: [],
        },
      ],
    })
  })

  it("should parse for loop with break statement", () => {
    const code = "for { break }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForStatement",
          label: undefined,
          loopId: "loop_1",
          body: [
            {
              kind: "BreakStatement",
              loopId: "loop_1",
            },
          ],
        },
      ],
    })
  })

  it("should not parse break statement outside of loop", () => {
    const code = "break"
    const tokens = tokenize("test", code)

    expect(() => new Parser(tokens).produceAST()).toThrow(
      "Unexpected break statement outside of loop."
    )
  })

  it("should parse nested for loop with break statment with correct loop id", () => {
    const code = "for { for { break } }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForStatement",
          label: undefined,
          loopId: "loop_1",
          body: [
            {
              kind: "ForStatement",
              label: undefined,
              loopId: "loop_2",
              body: [
                {
                  kind: "BreakStatement",
                  loopId: "loop_2",
                },
              ],
            },
          ],
        },
      ],
    })
  })

  it("should parse nested for loop with label and break statement with correct loop id", () => {
    const code = "for label outer { for label inner { break outer } break }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForStatement",
          label: "outer",
          loopId: "loop_1",
          body: [
            {
              kind: "ForStatement",
              label: "inner",
              loopId: "loop_2",
              body: [
                {
                  kind: "BreakStatement",
                  loopId: "loop_1",
                },
              ],
            },
            {
              kind: "BreakStatement",
              loopId: "loop_1",
            },
          ],
        },
      ],
    })
  })

  it("should parse for loop with continue statement", () => {
    const code = "for { skip }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForStatement",
          label: undefined,
          loopId: "loop_1",
          body: [
            {
              kind: "ContinueStatement",
              loopId: "loop_1",
            },
          ],
        },
      ],
    })
  })

  it("should not parse continue statement outside of loop", () => {
    const code = "skip"
    const tokens = tokenize("test", code)

    expect(() => new Parser(tokens).produceAST()).toThrow(
      "Unexpected continue statement outside of loop."
    )
  })

  it("should parse nested for loop with continue statment with correct loop id", () => {
    const code = "for { for { skip } }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForStatement",
          label: undefined,
          loopId: "loop_1",
          body: [
            {
              kind: "ForStatement",
              label: undefined,
              loopId: "loop_2",
              body: [
                {
                  kind: "ContinueStatement",
                  loopId: "loop_2",
                },
              ],
            },
          ],
        },
      ],
    })
  })

  it("should parse nested for loop with label and continue statement with correct loop id", () => {
    const code = "for label outer { for label inner { skip outer } skip }"
    const tokens = tokenize("test", code)

    const ast = new Parser(tokens).produceAST()

    expect(ast).toEqual({
      kind: "Program",
      body: [
        {
          kind: "ForStatement",
          label: "outer",
          loopId: "loop_1",
          body: [
            {
              kind: "ForStatement",
              label: "inner",
              loopId: "loop_2",
              body: [
                {
                  kind: "ContinueStatement",
                  loopId: "loop_1",
                },
              ],
            },
            {
              kind: "ContinueStatement",
              loopId: "loop_1",
            },
          ],
        },
      ],
    })
  })

  it("should not parse break statement with invalid label", () => {
    const code = "for { break invalid }"
    const tokens = tokenize("test", code)

    expect(() => new Parser(tokens).produceAST()).toThrow(
      "Invalid label 'invalid' for break statement."
    )
  })

  it("should not parse continue statement with invalid label", () => {
    const code = "for { skip invalid }"
    const tokens = tokenize("test", code)

    expect(() => new Parser(tokens).produceAST()).toThrow(
      "Invalid label 'invalid' for skip statement."
    )
  })
})
