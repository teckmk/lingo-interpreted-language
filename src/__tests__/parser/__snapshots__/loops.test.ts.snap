// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser - Loops should parse for in loop 1`] = `
{
  "body": [
    {
      "body": [],
      "indexIdentifier": {
        "position": {
          "end": {
            "column": 14,
            "line": 1,
          },
          "start": {
            "column": 13,
            "line": 1,
          },
        },
        "type": "IDENTIFIER",
        "value": "i",
      },
      "iterable": {
        "kind": "Identifier",
        "symbol": {
          "position": {
            "end": {
              "column": 23,
              "line": 1,
            },
            "start": {
              "column": 20,
              "line": 1,
            },
          },
          "value": "arr",
        },
      },
      "kind": "ForInStatement",
      "label": undefined,
      "loopId": "loop_1",
      "valueIdentifier": {
        "position": {
          "end": {
            "column": 16,
            "line": 1,
          },
          "start": {
            "column": 15,
            "line": 1,
          },
        },
        "type": "IDENTIFIER",
        "value": "v",
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - Loops should parse for in range loop with label 1`] = `
{
  "body": [
    {
      "body": [],
      "end": {
        "kind": "NumericLiteral",
        "value": {
          "position": {
            "end": {
              "column": 33,
              "line": 1,
            },
            "start": {
              "column": 31,
              "line": 1,
            },
          },
          "value": 10,
        },
      },
      "inclusive": false,
      "indexIdentifier": {
        "position": {
          "end": {
            "column": 14,
            "line": 1,
          },
          "start": {
            "column": 13,
            "line": 1,
          },
        },
        "type": "IDENTIFIER",
        "value": "i",
      },
      "kind": "ForRangeStatement",
      "label": {
        "position": {
          "end": {
            "column": 44,
            "line": 1,
          },
          "start": {
            "column": 40,
            "line": 1,
          },
        },
        "value": "loop",
      },
      "loopId": "loop_1",
      "start": {
        "kind": "NumericLiteral",
        "value": {
          "position": {
            "end": {
              "column": 27,
              "line": 1,
            },
            "start": {
              "column": 26,
              "line": 1,
            },
          },
          "value": 0,
        },
      },
      "step": undefined,
      "valueIdentifier": {
        "position": {
          "end": {
            "column": 16,
            "line": 1,
          },
          "start": {
            "column": 15,
            "line": 1,
          },
        },
        "type": "IDENTIFIER",
        "value": "v",
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - Loops should parse for loop with break statement 1`] = `
{
  "body": [
    {
      "body": [
        {
          "kind": "BreakStatement",
          "label": undefined,
          "loopId": "loop_1",
          "value": undefined,
        },
      ],
      "kind": "ForStatement",
      "label": undefined,
      "loopId": "loop_1",
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - Loops should parse for loop with condition 1`] = `
{
  "body": [
    {
      "body": [],
      "condition": {
        "kind": "BinaryExpr",
        "left": {
          "kind": "Identifier",
          "symbol": {
            "position": {
              "end": {
                "column": 6,
                "line": 1,
              },
              "start": {
                "column": 5,
                "line": 1,
              },
            },
            "value": "i",
          },
        },
        "operator": {
          "position": {
            "end": {
              "column": 8,
              "line": 1,
            },
            "start": {
              "column": 7,
              "line": 1,
            },
          },
          "value": "<",
        },
        "right": {
          "kind": "NumericLiteral",
          "value": {
            "position": {
              "end": {
                "column": 11,
                "line": 1,
              },
              "start": {
                "column": 9,
                "line": 1,
              },
            },
            "value": 10,
          },
        },
      },
      "kind": "ForStatement",
      "label": undefined,
      "loopId": "loop_1",
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - Loops should parse for loop with initializer, condition and update 1`] = `
{
  "body": [
    {
      "body": [],
      "condition": {
        "kind": "BinaryExpr",
        "left": {
          "kind": "Identifier",
          "symbol": {
            "position": {
              "end": {
                "column": 21,
                "line": 1,
              },
              "start": {
                "column": 20,
                "line": 1,
              },
            },
            "value": "i",
          },
        },
        "operator": {
          "position": {
            "end": {
              "column": 23,
              "line": 1,
            },
            "start": {
              "column": 22,
              "line": 1,
            },
          },
          "value": "<",
        },
        "right": {
          "kind": "NumericLiteral",
          "value": {
            "position": {
              "end": {
                "column": 26,
                "line": 1,
              },
              "start": {
                "column": 24,
                "line": 1,
              },
            },
            "value": 10,
          },
        },
      },
      "initializer": {
        "identifier": {
          "position": {
            "end": {
              "column": 14,
              "line": 1,
            },
            "start": {
              "column": 13,
              "line": 1,
            },
          },
          "value": "i",
        },
        "kind": "VarDeclaration",
        "modifier": "variable",
        "type": undefined,
        "value": {
          "kind": "NumericLiteral",
          "value": {
            "position": {
              "end": {
                "column": 18,
                "line": 1,
              },
              "start": {
                "column": 17,
                "line": 1,
              },
            },
            "value": 0,
          },
        },
      },
      "kind": "ForStatement",
      "label": undefined,
      "loopId": "loop_1",
      "update": {
        "assigne": {
          "kind": "Identifier",
          "symbol": {
            "position": {
              "end": {
                "column": 29,
                "line": 1,
              },
              "start": {
                "column": 28,
                "line": 1,
              },
            },
            "value": "i",
          },
        },
        "kind": "AssignmentExpr",
        "value": {
          "kind": "BinaryExpr",
          "left": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 33,
                  "line": 1,
                },
                "start": {
                  "column": 32,
                  "line": 1,
                },
              },
              "value": "i",
            },
          },
          "operator": {
            "position": {
              "end": {
                "column": 35,
                "line": 1,
              },
              "start": {
                "column": 34,
                "line": 1,
              },
            },
            "value": "+",
          },
          "right": {
            "kind": "NumericLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 37,
                  "line": 1,
                },
                "start": {
                  "column": 36,
                  "line": 1,
                },
              },
              "value": 1,
            },
          },
        },
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - Loops should parse for loop with label 1`] = `
{
  "body": [
    {
      "body": [],
      "kind": "ForStatement",
      "label": {
        "position": {
          "end": {
            "column": 15,
            "line": 1,
          },
          "start": {
            "column": 11,
            "line": 1,
          },
        },
        "value": "loop",
      },
      "loopId": "loop_1",
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - Loops should parse for loop with label and condition 1`] = `
{
  "body": [
    {
      "body": [],
      "condition": {
        "kind": "BinaryExpr",
        "left": {
          "kind": "Identifier",
          "symbol": {
            "position": {
              "end": {
                "column": 6,
                "line": 1,
              },
              "start": {
                "column": 5,
                "line": 1,
              },
            },
            "value": "i",
          },
        },
        "operator": {
          "position": {
            "end": {
              "column": 8,
              "line": 1,
            },
            "start": {
              "column": 7,
              "line": 1,
            },
          },
          "value": "<",
        },
        "right": {
          "kind": "NumericLiteral",
          "value": {
            "position": {
              "end": {
                "column": 11,
                "line": 1,
              },
              "start": {
                "column": 9,
                "line": 1,
              },
            },
            "value": 10,
          },
        },
      },
      "kind": "ForStatement",
      "label": {
        "position": {
          "end": {
            "column": 22,
            "line": 1,
          },
          "start": {
            "column": 18,
            "line": 1,
          },
        },
        "value": "loop",
      },
      "loopId": "loop_1",
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - Loops should parse for loop with label, initializer, condition and update 1`] = `
{
  "body": [
    {
      "body": [],
      "condition": {
        "kind": "BinaryExpr",
        "left": {
          "kind": "Identifier",
          "symbol": {
            "position": {
              "end": {
                "column": 21,
                "line": 1,
              },
              "start": {
                "column": 20,
                "line": 1,
              },
            },
            "value": "i",
          },
        },
        "operator": {
          "position": {
            "end": {
              "column": 23,
              "line": 1,
            },
            "start": {
              "column": 22,
              "line": 1,
            },
          },
          "value": "<",
        },
        "right": {
          "kind": "NumericLiteral",
          "value": {
            "position": {
              "end": {
                "column": 26,
                "line": 1,
              },
              "start": {
                "column": 24,
                "line": 1,
              },
            },
            "value": 10,
          },
        },
      },
      "initializer": {
        "identifier": {
          "position": {
            "end": {
              "column": 14,
              "line": 1,
            },
            "start": {
              "column": 13,
              "line": 1,
            },
          },
          "value": "i",
        },
        "kind": "VarDeclaration",
        "modifier": "variable",
        "type": undefined,
        "value": {
          "kind": "NumericLiteral",
          "value": {
            "position": {
              "end": {
                "column": 18,
                "line": 1,
              },
              "start": {
                "column": 17,
                "line": 1,
              },
            },
            "value": 0,
          },
        },
      },
      "kind": "ForStatement",
      "label": {
        "position": {
          "end": {
            "column": 48,
            "line": 1,
          },
          "start": {
            "column": 44,
            "line": 1,
          },
        },
        "value": "loop",
      },
      "loopId": "loop_1",
      "update": {
        "assigne": {
          "kind": "Identifier",
          "symbol": {
            "position": {
              "end": {
                "column": 29,
                "line": 1,
              },
              "start": {
                "column": 28,
                "line": 1,
              },
            },
            "value": "i",
          },
        },
        "kind": "AssignmentExpr",
        "value": {
          "kind": "BinaryExpr",
          "left": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 33,
                  "line": 1,
                },
                "start": {
                  "column": 32,
                  "line": 1,
                },
              },
              "value": "i",
            },
          },
          "operator": {
            "position": {
              "end": {
                "column": 35,
                "line": 1,
              },
              "start": {
                "column": 34,
                "line": 1,
              },
            },
            "value": "+",
          },
          "right": {
            "kind": "NumericLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 37,
                  "line": 1,
                },
                "start": {
                  "column": 36,
                  "line": 1,
                },
              },
              "value": 1,
            },
          },
        },
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - Loops should parse for range loop 1`] = `
{
  "body": [
    {
      "body": [],
      "end": {
        "kind": "NumericLiteral",
        "value": {
          "position": {
            "end": {
              "column": 33,
              "line": 1,
            },
            "start": {
              "column": 31,
              "line": 1,
            },
          },
          "value": 10,
        },
      },
      "inclusive": false,
      "indexIdentifier": {
        "position": {
          "end": {
            "column": 14,
            "line": 1,
          },
          "start": {
            "column": 13,
            "line": 1,
          },
        },
        "type": "IDENTIFIER",
        "value": "i",
      },
      "kind": "ForRangeStatement",
      "label": undefined,
      "loopId": "loop_1",
      "start": {
        "kind": "NumericLiteral",
        "value": {
          "position": {
            "end": {
              "column": 27,
              "line": 1,
            },
            "start": {
              "column": 26,
              "line": 1,
            },
          },
          "value": 0,
        },
      },
      "step": undefined,
      "valueIdentifier": {
        "position": {
          "end": {
            "column": 16,
            "line": 1,
          },
          "start": {
            "column": 15,
            "line": 1,
          },
        },
        "type": "IDENTIFIER",
        "value": "v",
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - Loops should parse for range loop with step 1`] = `
{
  "body": [
    {
      "body": [],
      "end": {
        "kind": "NumericLiteral",
        "value": {
          "position": {
            "end": {
              "column": 38,
              "line": 1,
            },
            "start": {
              "column": 36,
              "line": 1,
            },
          },
          "value": 10,
        },
      },
      "inclusive": true,
      "indexIdentifier": {
        "position": {
          "end": {
            "column": 14,
            "line": 1,
          },
          "start": {
            "column": 13,
            "line": 1,
          },
        },
        "type": "IDENTIFIER",
        "value": "i",
      },
      "kind": "ForRangeStatement",
      "label": undefined,
      "loopId": "loop_1",
      "start": {
        "kind": "NumericLiteral",
        "value": {
          "position": {
            "end": {
              "column": 27,
              "line": 1,
            },
            "start": {
              "column": 26,
              "line": 1,
            },
          },
          "value": 0,
        },
      },
      "step": {
        "kind": "NumericLiteral",
        "value": {
          "position": {
            "end": {
              "column": 45,
              "line": 1,
            },
            "start": {
              "column": 44,
              "line": 1,
            },
          },
          "value": 2,
        },
      },
      "valueIdentifier": {
        "position": {
          "end": {
            "column": 16,
            "line": 1,
          },
          "start": {
            "column": 15,
            "line": 1,
          },
        },
        "type": "IDENTIFIER",
        "value": "v",
      },
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - Loops should parse infinite for loop 1`] = `
{
  "body": [
    {
      "body": [],
      "kind": "ForStatement",
      "label": undefined,
      "loopId": "loop_1",
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - Loops should parse nested for loop with break statement with correct loop id 1`] = `
{
  "body": [
    {
      "body": [
        {
          "body": [
            {
              "kind": "BreakStatement",
              "label": undefined,
              "loopId": "loop_2",
              "value": undefined,
            },
          ],
          "kind": "ForStatement",
          "label": undefined,
          "loopId": "loop_2",
        },
      ],
      "kind": "ForStatement",
      "label": undefined,
      "loopId": "loop_1",
    },
  ],
  "kind": "Program",
}
`;

exports[`Parser - Loops should parse nested for loop with label and break statement with correct loop id 1`] = `
{
  "body": [
    {
      "body": [
        {
          "body": [
            {
              "kind": "BreakStatement",
              "label": {
                "position": {
                  "end": {
                    "column": 48,
                    "line": 1,
                  },
                  "start": {
                    "column": 43,
                    "line": 1,
                  },
                },
                "value": "outer",
              },
              "loopId": "loop_1",
              "value": undefined,
            },
          ],
          "kind": "ForStatement",
          "label": {
            "position": {
              "end": {
                "column": 34,
                "line": 1,
              },
              "start": {
                "column": 29,
                "line": 1,
              },
            },
            "value": "inner",
          },
          "loopId": "loop_2",
        },
        {
          "kind": "BreakStatement",
          "label": undefined,
          "loopId": "loop_1",
          "value": undefined,
        },
      ],
      "kind": "ForStatement",
      "label": {
        "position": {
          "end": {
            "column": 16,
            "line": 1,
          },
          "start": {
            "column": 11,
            "line": 1,
          },
        },
        "value": "outer",
      },
      "loopId": "loop_1",
    },
  ],
  "kind": "Program",
}
`;
